File: web/src/lib/weightFinder.ts
// Deterministic weight-finder rule engine.
// Spec reference: docs/weight-finder-rules.md, section D.2.
//
// This is the browser-side implementation. It is kept in lockstep with the C# reference
// implementation at src/KettlebellFinder.Domain/WeightFinder.cs — same steps, same order,
// same D.2 test table. Change one, change the other.
//
// LIABILITY (§7): output is always a range plus an edge hint, never a bare number.
// No body metrics are accepted, computed, or stored — every input below is non-metric.
// Nothing is persisted; this runs entirely client-side (spec D).

export type Focus = "strength_press" | "mixed" | "conditioning_swings";

export type TrainingBackground =
  | "untrained"
  | "occasional"
  | "active_no_strength"
  | "strength_trained";

export type Sex = "female" | "male";

export type AgeBand = "under_50" | "age_50_plus";

/** Where within the recommended range to start. Advisory only — never collapses the range. */
export type Edge = "lower" | "upper" | "either";

/** Bodyweight self-check plus an optional press proxy. All non-metric (§7). */
export interface TechniqueCheck {
  squatClean: boolean;
  hingeClean: boolean;
  overheadClean: boolean;
  /** Optional 5 kg overhead-press proxy — press only, never a swing. Defaults to false. */
  pressProxyClean?: boolean;
}

export interface WeightFinderInput {
  focus: Focus;
  trainingBackground: TrainingBackground;
  technique: TechniqueCheck;
  sex?: Sex;
  ageBand?: AgeBand;
}

/** Recommended weight range in kg. Always a span (§7). */
export interface WeightRange {
  lowKg: number;
  highKg: number;
}

export interface WeightRecommendation {
  range: WeightRange;
  edge: Edge;
}

// Ordered by training background — index matches BAND_ORDER below.
const BAND_ORDER: TrainingBackground[] = [
  "untrained",
  "occasional",
  "active_no_strength",
  "strength_trained",
];

// Base ranges per training background (spec D.1). WORKING VALUES: the four-band structure
// is stable, the exact kg boundaries are not frozen and must be checked against sources
// before go-live (spec F.1). Indexed like BAND_ORDER.
const BANDS: readonly WeightRange[] = [
  { lowKg: 8, highKg: 10 }, // untrained
  { lowKg: 10, highKg: 12 }, // occasional
  { lowKg: 12, highKg: 14 }, // active_no_strength
  { lowKg: 14, highKg: 18 }, // strength_trained
];

const clamp = (value: number, min: number, max: number): number =>
  Math.min(Math.max(value, min), max);

/**
 * Resolves an input to exactly one recommendation by applying the spec D.2 steps in order.
 * Pure and deterministic: the same input always yields the same output.
 */
export function resolveWeight(input: WeightFinderInput): WeightRecommendation {
  const technique = input.technique;

  const baseBand = BAND_ORDER.indexOf(input.trainingBackground);
  let band = baseBand;
  let edge: Edge = "either";

  // Step 2 — focus.
  switch (input.focus) {
    case "strength_press":
      edge = "lower";
      break;
    case "conditioning_swings":
      edge = "upper";
      // Only bump the band up on a demonstrably clean overhead press.
      if (technique.overheadClean && technique.pressProxyClean) {
        band += 1;
      }
      break;
    case "mixed":
    default:
      break;
  }

  // Step 3 — technique. Any mandatory movement not clean pulls to the lower edge,
  // overriding an upper edge set by a conditioning focus.
  const techniqueClean =
    technique.squatClean && technique.hingeClean && technique.overheadClean;
  if (!techniqueClean) {
    edge = "lower";
  }

  // Step 4 — sex (optional). The press ceiling limits a single all-purpose bell more,
  // so shift one band down. Range modifier only (§7).
  if (input.sex === "female") {
    band -= 1;
  }

  // Step 5 — age (optional). Only relevant for an untrained starter.
  if (input.ageBand === "age_50_plus" && input.trainingBackground === "untrained") {
    edge = "lower";
  }

  // Step 6 — cap the net band shift to one band from base, then clamp to the valid range.
  band = clamp(band, baseBand - 1, baseBand + 1);
  band = clamp(band, 0, BANDS.length - 1);

  return { range: BANDS[band], edge };
}

File: src/KettlebellFinder.Domain/WeightFinder.cs
namespace KettlebellFinder.Domain;

// Deterministic weight-finder rule engine.
// Spec reference: weight-finder-rules.md, section D.2.
//
// LIABILITY (§7): output is always a range plus an edge hint, never a bare number.
// No body metrics are accepted, computed, or stored — every input below is non-metric.

public enum Focus
{
    StrengthPress,
    Mixed,
    ConditioningSwings,
}

public enum TrainingBackground
{
    Untrained = 0,
    Occasional = 1,
    ActiveNoStrength = 2,
    StrengthTrained = 3,
}

public enum Sex
{
    Female,
    Male,
}

public enum AgeBand
{
    Under50,
    Age50Plus,
}

/// <summary>Where within the recommended range to start. Advisory text only — never collapses the range.</summary>
public enum Edge
{
    Lower,
    Upper,
    Either,
}

/// <summary>Bodyweight self-check plus an optional press proxy. All non-metric (§7).</summary>
public sealed record TechniqueCheck(
    bool SquatClean,
    bool HingeClean,
    bool OverheadClean,
    bool PressProxyClean = false);

public sealed record WeightFinderInput(
    Focus Focus,
    TrainingBackground TrainingBackground,
    TechniqueCheck Technique,
    Sex? Sex = null,
    AgeBand? AgeBand = null);

/// <summary>Recommended weight range in kg. Always a span (§7).</summary>
public sealed record WeightRange(int LowKg, int HighKg);

public sealed record WeightRecommendation(WeightRange Range, Edge Edge);

public static class WeightFinder
{
    // Base ranges per training background (spec D.1). WORKING VALUES: the four-band structure
    // is stable, the exact kg boundaries are not frozen and must be checked against sources
    // before go-live (spec F.1). Indexed by TrainingBackground.
    private static readonly WeightRange[] Bands =
    {
        new(8, 10),   // Untrained
        new(10, 12),  // Occasional
        new(12, 14),  // ActiveNoStrength
        new(14, 18),  // StrengthTrained
    };

    /// <summary>
    /// Resolves an input to exactly one recommendation by applying the spec D.2 steps in order.
    /// Pure and deterministic: the same input always yields the same output.
    /// </summary>
    public static WeightRecommendation Resolve(WeightFinderInput input)
    {
        ArgumentNullException.ThrowIfNull(input);
        ArgumentNullException.ThrowIfNull(input.Technique);

        int baseBand = (int)input.TrainingBackground;
        int band = baseBand;
        var edge = Edge.Either;

        // Step 2 — focus.
        switch (input.Focus)
        {
            case Focus.StrengthPress:
                edge = Edge.Lower;
                break;
            case Focus.ConditioningSwings:
                edge = Edge.Upper;
                // Only bump the band up on a demonstrably clean overhead press.
                if (input.Technique.OverheadClean && input.Technique.PressProxyClean)
                {
                    band += 1;
                }
                break;
            case Focus.Mixed:
            default:
                break;
        }

        // Step 3 — technique. Any mandatory movement not clean pulls to the lower edge,
        // overriding an upper edge set by a conditioning focus.
        bool techniqueClean =
            input.Technique.SquatClean
            && input.Technique.HingeClean
            && input.Technique.OverheadClean;
        if (!techniqueClean)
        {
            edge = Edge.Lower;
        }

        // Step 4 — sex (optional). The press ceiling limits a single all-purpose bell more,
        // so shift one band down. Range modifier only (§7).
        if (input.Sex == Sex.Female)
        {
            band -= 1;
        }

        // Step 5 — age (optional). Only relevant for an untrained starter.
        if (input.AgeBand == AgeBand.Age50Plus
            && input.TrainingBackground == TrainingBackground.Untrained)
        {
            edge = Edge.Lower;
        }

        // Step 6 — cap the net band shift to one band from base, then clamp to the valid range.
        band = Math.Clamp(band, baseBand - 1, baseBand + 1);
        band = Math.Clamp(band, 0, Bands.Length - 1);

        return new WeightRecommendation(Bands[band], edge);
    }
}
